platforms/cuda/src/CudaKernels.cpp

```diff
@@ -1079,18 +1079,32 @@ void CudaCalcNonbondedForceKernel::copyParametersToContext(ContextImpl& context,
     
     // Record the per-particle parameters.
     
-    vector<float4> baseParticleParamVec(cu.getPaddedNumAtoms(), make_float4(0, 0, 0, 0));
-    const vector<int>& order = cu.getAtomIndex();
-    for (int i = 0; i < force.getNumParticles(); i++) {
-        double charge, sigma, epsilon;
-        force.getParticleParameters(i, charge, sigma, epsilon);
-        baseParticleParamVec[i] = make_float4(charge, sigma, epsilon, 0);
+    if (firstParticle <= lastParticle) {
+        vector<float4> baseParticleParamVec(cu.getPaddedNumAtoms(), make_float4(0, 0, 0, 0));
+        for (int i = 0; i < force.getNumParticles(); i++) {
+            double charge, sigma, epsilon;
+            force.getParticleParameters(i, charge, sigma, epsilon);
+            baseParticleParamVec[i] = make_float4(charge, sigma, epsilon, 0);
+        }
+        baseParticleParams.uploadSubArray(&baseParticleParamVec[firstParticle], firstParticle, lastParticle-firstParticle+1);
+
+        // Compute the self energy.
+
+        ewaldSelfEnergy = 0.0;
+        if (nonbondedMethod == Ewald || nonbondedMethod == PME || nonbondedMethod == LJPME) {
+            if (cu.getContextIndex() == 0) {
+                for (int i = 0; i < force.getNumParticles(); i++) {
+                    ewaldSelfEnergy -= baseParticleParamVec[i].x*baseParticleParamVec[i].x*ONE_4PI_EPS0*alpha/sqrt(M_PI);
+                    if (doLJPME)
+                        ewaldSelfEnergy += baseParticleParamVec[i].z*pow(baseParticleParamVec[i].y*dispersionAlpha, 6)/3.0;
+                }
+            }
+        }
```

Comment:
> but was not able to see any significant differences between the two codes

At first glance this could be the main difference with #4577 

The whole `baseParticleParamVec` is allocated and filled, then only a part of it used for uploading but the whole vector  - to calculate energies if needed.

On the other hand #4577 creates a small vector only for uploading and calculates energies by calling force.getParticleParameters for all atoms.

Considering that the allocation and filling of the whole vector takes time, but the benchmark (done by @akalpokas) uses RF (`(nonbondedMethod == Ewald || nonbondedMethod == PME || nonbondedMethod == LJPME)` is false, right?), this time is wasted.

Reply:
Be aware that #4577 has a bug: it doesn't revalidate the exceptions for a NonbondedForce.  I suspect that's where a lot of its apparent speedup comes from.  At the end of `copySomeParametersToContext()` where it calls `invalidateMoleculesByAtom()`, add a call to `invalidateMoleculesByGroup()` immediately after it.  How does that affect the above numbers?

Reply:
I have reran the benchmark with your suggested additions and the performance between the two implementations is now essentially identical (please see my updated original comment) - initial addition of `invalidateMoleculesByGroup()` to `platforms/cuda/src/CudaKernels.cpp` only regressed the performance of 4577 by around 3%.  Once I updated `platforms/common/src/CommonKernels.cpp` as well, the performance between both implementations essentially equalized.